docs/doc/90-contributing/09-how-to-write-aggregate-functions.md

```diff
@@ -2,135 +2,126 @@
 
+The main functions are:
 
-- The function `accumulate_keys` is similar to accumulate, but we must take into consideration the keys and offsets, for which each key represents a unique memory address named place.
-- The function `serialize` serializes state into binary.
-- The function `deserialize` deserializes state from binary.
-- The function `merge`, can be used to merge other state into current state.
-- The function `merge_result`, can be used to represent the aggregate function state into one-row field.
-
+- **name** represents the name of the function being registered, such as avg, sum, etc.
+- **return_type** represents the return type of the registered function. The return value of the same function may change due to different parameter types. For example, the parameter of `sum(int8)` is of type `i8`, but the return value may be `int64`.
+- **init_state** is used to initialize the state of the aggregate function.
+- **state_layout** is used to represent the size of the **state** in memory and the arrangement of memory blocks.
+- **accumulate** is used for `SingleStateAggregator`. That is, the entire block can be aggregated under a single state without any keys. For example, when `selecting count(*) from t`, there is no grouping column in the query, and the `accumulate` function will be called.
+- **accumulate_keys** is used for `PartialAggregator`. Here, `key` and `offset` need to be considered, where each key represents a unique memory address, denoted as the function parameter place.
+- **serialize** serializes the **state** in the aggregation process into binary.
+- **deserialize** deserializes the binary into **state**.
+- **merge** is used to* merge other state into the current state.
+- **merge_result** can merge the **Aggregate Function state** into a single value.
 
 ## Example
-Let's take an example of aggregate function `sum`.
 
-It's declared as `AggregateSumFunction<T, SumT>`, we can accept varying integer types like `u8`, `i8`. `T` and `SumT` is logic types which implement `DFPrimitiveType`. e.g., `T` is `u8` and `SumT` must be `u64`.
+**Take avg as an example**
+
+The implementation details can be found in [aggregate_avg.rs](https://github.com/datafuselabs/databend/blob/d5e06af03ba0f99afdd6bdc974bf2f5c1c022db8/src/query/functions/src/aggregates/aggregate\_avg.rs).
 
-Also, we can dispatch it using macros by matching the types of the arguments. Take a look at the `with_match_primitive_type` to understand the dispatch macros.
+Because we need to accumulate each value and divide by the total number of non-null rows. Therefore, the avg function is defined as the `struct AggregateAvgFunction<T, SumT>`. Where `T` and `SumT` are logical types that implement [Number](https://github.com/datafuselabs/databend/blob/2aec38605eebb7f0e1717f7f54ec52ae0f2e530b/src/query/expression/src/types/number.rs).
```

Comment:
Because we need to accumulate each value and divide them by the total number of non-null rows, the avg function is defined as a `struct AggregateAvgFunction<T, SumT>` where `T` and `SumT` are logical types that implement [Number]........